### PR TITLE
Teach fossil to use HTTP Basic and Fossil built-in authentication

### PR DESCRIPTION
**Pull request by kthbit**
Composer was unable to clone/pull from a Fossil repo that was protected by either HTTP Basic or Fossil  authentication. This patch attempts to fix that. Both (or either) authentications are stored in auth.json. HTTP Basic auth is specified as before, per domain. Fossil credentials are specified per remote repository, in case pulling from multiple remote repositories with different credentials for each.

### CODE REVIEW

File: src/Composer/Downloader/FossilDownloader.php [hunk 1]
```diff
@@ -113,4 +121,108 @@ protected function hasMetadataRepository($path)
...
+
+        // Only for http(s) URLs.
+        if (!isset($parsed['scheme']) || ($parsed['scheme'] != 'http' && $parsed['scheme'] != 'https') || empty($parsed['host']))
+            return '';
+
+        $httpBasicConfig = $this->config->get('http-basic');
+        if (isset($httpBasicConfig[$parsed['host']])) {
+            $creds = $httpBasicConfig[$parsed['host']];
+            // We don't allow empty username or password. Both must be defined.
+            if (!empty($creds['username']) && !empty($creds['password']))
```
Comment (stof) [inline src/Composer/Downloader/FossilDownloader.php]: please add the curly braces. Not having them makes it much easier to introduce a mistake later (not realizing that the second line is not inside the if even though you intended it to be)

File: src/Composer/Repository/Vcs/FossilDriver.php [hunk 2]
```diff
@@ -79,24 +79,28 @@ protected function updateLocalRepo()
...
-        if (is_file($this->repoFile) && is_dir($this->checkoutDir) && 0 === $this->process->execute('fossil info', $output, $this->checkoutDir)) {
-            if (0 !== $this->process->execute('fossil pull', $output, $this->checkoutDir)) {
+        if ($this->isLocalCheckout()) {
+            $cmd = $this->getPullCommand();
+            if (0 !== $this->process->execute($cmd, $output, $this->checkoutDir)) {
                 $this->io->writeError('<error>Failed to update '.$this->url.', package information from this repository may be outdated ('.$this->process->getErrorOutput().')</error>');
             }
-        } else {
+        }
+        else {
```
Comment (stof) [inline src/Composer/Repository/Vcs/FossilDriver.php]: please don't change the coding standards

File: src/Composer/Downloader/FossilDownloader.php (hunk 1 above)
Comment (stof) [reply src/Composer/Downloader/FossilDownloader.php]: this applies to the whole PR btw. I won't comment on all of them.

File: src/Composer/Downloader/FossilDownloader.php [hunk 3]
```diff
@@ -113,4 +121,117 @@ protected function hasMetadataRepository($path)
...
+     *
+     * It is expected that credentials specific for each repository would be located in auth.json in the following structure:
+     *
+     * {
+     *   "fossil-auth": {
+     *     "host.com/path/to/repo": {
+     *       "username": "my-user",
+     *       "password": "my-pass"
+     *     }
+     *   }
```
Comment (Seldaek) [inline src/Composer/Downloader/FossilDownloader.php]: Is it not possible to simply reuse the existing [http-basic](https://getcomposer.org/doc/06-config.md#http-basic) key for this and do it only based on hostname? Especially as you seem to have both http basic + fossil level auth, calling the http basic one fossil is quite confusing.

File: src/Composer/Repository/Vcs/FossilDriver.php [hunk 4]
```diff
@@ -90,20 +91,159 @@ protected function updateLocalRepo()
...
+                $parsed['host'],
+                isset($parsed['port']) ? ':'.$parsed['port'] : '',
+                $parsed['path'],
+                isset($parsed['query']) ? '?'.$parsed['query'] : '',
+                isset($parsed['fragment']) ? '#'.$parsed['fragment'] : ''
+            );
+        }
+
+        return $url;
+    }
```
Comment (Seldaek) [inline src/Composer/Repository/Vcs/FossilDriver.php]: getHttpBasicAuth and getUrlWithAuth seem to be pretty much duplicated from the FossilDriver.. It'd be good to instead have a Util class to deduplicate the code.

**Comment by Seldaek** [comment]
Closing for now, happy to receive an updated PR with the changes above.

**Pull request closed without merging**